File: config/src/test/scala/net/kemitix/thorp/config/ConfigOptionTest.scala
package net.kemitix.thorp.config

import net.kemitix.thorp.domain.Sources
import net.kemitix.thorp.filesystem.TemporaryFolder
import org.scalatest.FunSpec
import zio.DefaultRuntime

import scala.jdk.CollectionConverters._

class ConfigOptionTest extends FunSpec with TemporaryFolder {

  describe("when more than one source") {
    it("should preserve their order") {
      withDirectory(path1 => {
        withDirectory(path2 => {
          val configOptions = ConfigOptions.create(
            List[ConfigOption](
              ConfigOption.source(path1),
              ConfigOption.source(path2),
              ConfigOption.bucket("bucket"),
              ConfigOption.ignoreGlobalOptions(),
              ConfigOption.ignoreUserOptions()
            ).asJava)
          val expected = Sources.create(List(path1, path2).asJava)
          val result   = invoke(configOptions)
          assert(result.isRight, result)
          assertResult(expected)(ConfigQuery.sources(configOptions))
        })
      })
    }
  }

  private def invoke(configOptions: ConfigOptions) = {
    new DefaultRuntime {}.unsafeRunSync {
      ConfigurationBuilder.buildConfig(configOptions)
    }.toEither
  }
}

File: config/src/test/java/net/kemitix/thorp/config/ConfigOptionTest.java
package net.kemitix.thorp.config;

import net.kemitix.thorp.domain.Sources;
import net.kemitix.thorp.filesystem.TemporaryFolder;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;

public class ConfigOptionTest
        implements TemporaryFolder, WithAssertions {
    @Test
    @DisplayName("when more the one source then preserve their order")
    public void whenMultiSource_PreserveOrder() {
        withDirectory(path1 -> {
            withDirectory(path2 -> {
                ConfigOptions configOptions = ConfigOptions.create(
                        Arrays.asList(
                                ConfigOption.source(path1),
                                ConfigOption.source(path2),
                                ConfigOption.bucket("bucket"),
                                ConfigOption.ignoreGlobalOptions(),
                                ConfigOption.ignoreUserOptions()
                        ));
                List<Path> expected = Arrays.asList(path1, path2);
                assertThatCode(() -> {
                    Configuration result =
                            ConfigurationBuilder.buildConfig(configOptions);
                    assertThat(result.sources.paths()).isEqualTo(expected);
                }).doesNotThrowAnyException();
            });
        });
    }
}
